### PR TITLE
(PUP-4201) Add support for structured logging

### PR DESCRIPTION
**Pull request by thallgren**
This pull-request removes the responsibility for formatting the log messages
from the `IssueReporter` and the `ExternalFilerError` classes. Instead,
the formatting is deferred and performed by the `Log.to_s` method (if
such formatting is at all needed).

Some instance variables were added, both to the `ExternalFileError` class
and to the Log class to accommodate detailed information about
issue_code, environment, node, and position.

A new Log destination _'json_file'_ was added. It will not use the
formatted message but instead output all details in a JSON style
Map. Exception backtrace is output as a JSON Array.

### CODE REVIEW

**Comment by cprice404** [comment]
Has this been tested with Puppet Server, or has any design work been put into the compatibility with the Puppet Server logging framework?

**Comment by thallgren** [comment]
It has not been tested with Puppet Server but care has been taken to not change the current log output.

File: lib/puppet/util/log.rb [hunk 1]
```diff
@@ -238,44 +249,41 @@ def self.validlevel?(level)
...
   def self.from_data_hash(data)
-    obj = allocate
-    obj.initialize_from_hash(data)
-    obj
+    symkeyed_data = {}
+    data.each_pair { |k,v| symkeyed_data[k.to_sym] = v }
+    new(symkeyed_data)
   end
 
-  attr_accessor :time, :remote, :file, :line, :source
```
Comment (hlindberg) [inline lib/puppet/util/log.rb]: was :remote never used? (it is missing in the new version).

File: lib/puppet/indirector/catalog/compiler.rb [hunk 2]
```diff
@@ -89,15 +89,9 @@ def compile(node)
 
     benchmark(:notice, str) do
       Puppet::Util::Profiler.profile(str, [:compiler, :compile, node.environment, node.name]) do
-        begin
-          config = Puppet::Parser::Compiler.compile(node)
-        rescue Puppet::Error => detail
-          Puppet.err(detail.to_s) if networked?
-          raise
-        end
```
Comment (hlindberg) [inline lib/puppet/indirector/catalog/compiler.rb]: where is the above done now? (seems like it logged only if networked (an agent request? or something else?). understand that new version does not catch anything (better), but want to make sure that the implication of not logging if networking is clear.

File: lib/puppet/util/log.rb [hunk 3]
```diff
@@ -18,11 +18,11 @@ class Puppet::Util::Log
   @desttypes = {}
 
   # Create a new destination type.
-  def self.newdesttype(name, options = {}, &block)
+  def self.newdesttype(name, options = {}, parent = nil, &block)
```
Comment (hlindberg) [inline lib/puppet/util/log.rb]: so this should be backwards compatible so this logdest works: https://github.com/puppetlabs/puppet-server/blob/master/src/ruby/puppet-server-lib/puppet/server/logger.rb

**Comment by hlindberg** [comment]
this should target the branch 3.8, not master

File: lib/puppet/util/log.rb (hunk 1 above)
Comment (thallgren) [inline lib/puppet/util/log.rb]: I couldn't find any usage of it anywhere in the code.

File: lib/puppet/indirector/catalog/compiler.rb (hunk 2 above)
Comment (thallgren) [inline lib/puppet/indirector/catalog/compiler.rb]: The only statement within begin/rescue here was a call to `Puppet::Parser::Compile.compile(node)` (the class method). That method will _always_ rescue, log, and re-raise all exceptions. So what this change does is avoiding a duplicate from being logged when networked.

File: lib/puppet/util/log.rb (hunk 3 above)
Comment (thallgren) [inline lib/puppet/util/log.rb]: Yes, it's backward compatible. The only dest_type that make use of this argument at present is the _json_file_ (passes the _file_ as it's parent). All others remain unchanged.

**Comment by thallgren** [comment]
This PR is replaced by [PR-3737](3737) that targets branch 3.8.

**Pull request closed without merging**